Olá, Gabriel

Meu nome é Ramon. Ao ao longo do texto farei algumas observações sobre melhorias no código. Nas partes em que não houver comentários, está tudo certo, ok? Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser consertado, caso contrário seu projeto **não** será aceito.
</div>

Você pode me responder usando isso:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

###

<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho.

Como pontos fortes do projeto, destaco:
- Código simples e eficiente (sem loop desnecessário)
- Uso adequado dos recursos da lição
- Calibração de hiperparâmetros está correta


Bom trabalho. Parabéns pelo empenho. Desejo sucesso na jornada.


</div>



In [1]:
#Importing all relevant libraries
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [2]:
#Transformando o dataset em Dataframe:
df = pd.read_csv('users_behavior.csv')

In [3]:
#Principais dados do DF:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB




<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

- É importante fazer etapa EDA: há outliers, a variável a ser predita está desbalanceada
</div>

In [4]:
#Dataframe dividido em três: treinamento, validação e teste  (60:20:20)
df_train, df_valid = train_test_split(df, random_state=54321, test_size = 0.2)
df_train, df_test = train_test_split(df_train, random_state=54321, test_size = 0.25)

In [5]:
#Verificando se o processo de split dos dados funcionou corretamente:
print(df.shape)
print(df_train.shape)
print(df_valid.shape)
print(df_test.shape)

(3214, 5)
(1928, 5)
(643, 5)
(643, 5)


In [6]:
#Criação das variáveis Característica e Objetivo:
train_features = df_train.drop(['is_ultra'], axis = 1)
train_target = df_train['is_ultra']
valid_features = df_valid.drop(['is_ultra'], axis = 1)
valid_target = df_valid['is_ultra']
test_features = df_test.drop(['is_ultra'], axis = 1)
test_target = df_test['is_ultra']

#### - Logistic Regression analysis

In [7]:
#Definição do modelo LR e cálculo do acurácia para o modelo (01):
model_LR1 = LogisticRegression(random_state=54321, solver='liblinear')
model_LR1.fit(train_features, train_target)
score_valid = model_LR1.score(valid_features, valid_target)

print("Acurácia do modelo de regressão logística no conjunto de validação:", score_valid)

Acurácia do modelo de regressão logística no conjunto de validação: 0.6780715396578538


In [8]:
#Definição do modelo LR e cálculo do acurácia para o modelo (02):
model_LR2 = LogisticRegression(random_state=54321, solver='lbfgs')
model_LR2.fit(train_features, train_target)
score_valid = model_LR2.score(valid_features, valid_target)

print("Acurácia do modelo de regressão logística no conjunto de validação:", score_valid)

Acurácia do modelo de regressão logística no conjunto de validação: 0.6734059097978227


In [9]:
#Definição do modelo LR e cálculo do acurácia para o modelo (03):
model_LR3 = LogisticRegression(random_state=54321, solver='newton-cg')
model_LR3.fit(train_features, train_target)
score_valid = model_LR3.score(valid_features, valid_target)

print("Acurácia do modelo de regressão logística no conjunto de validação:", score_valid)

Acurácia do modelo de regressão logística no conjunto de validação: 0.7076205287713841


#### - Random Forest Classifier analysis

In [10]:
best_score = 0
best_est = 0
#Criação de loop para análise do hiperparâmetro n_etimators:
for est in range(10, 81, 10):
#Definição do modelo RFC e cálculo de acurácia para o modelo:
    model_RFC = RandomForestClassifier(random_state=54321, n_estimators=est)
    model_RFC.fit(train_features, train_target)
    score = model_RFC.score(valid_features, valid_target)
    if score > best_score:
        best_score = score
        best_est = est

print(f"Acurácia do melhor modelo no conjunto de validação (n_estimators = {best_est}): {best_score}".format(best_est, best_score))

Acurácia do melhor modelo no conjunto de validação (n_estimators = 40): 0.7947122861586314


- #### Resultados da etapa de treinamento e validação dos dados

Como pudemos ver, avaliamos a aplicabilidade de dois modelos de classificação distintos:
 - Para o modelo Logistic Regression, permutamos o hiperparâmetro "solver" para avaliar o que melhor se encaixaria no nosso problema. Dentre os três avaliados, que são em geral mais utilizados para dados de tamanho pequeno e médio (que é o nosso caso), o modelo com solver = 'newton-cg' melhor se adequou aos dados de validação, chegando a 70% de acurácia.
 - Para o modelo Random Forest Classifier, variamos o hiperparâmetro n_estimators em loop variando de 10 em 10 até 80. Chegamos a conclusão que o mais adequado ao nosso problema possui hiperparâmetro = 40, e para o conjunto de validação obteve um bom resultado de acurácia, atingindo quase 80%.

- #### Acurácia do conjunto de teste

In [11]:
#Acurácia do modelo Logistic Regression:
score_lr = model_LR3.score(test_features, test_target)
#Acurácia do modelo Random Forest Classifier:
model_RFC = RandomForestClassifier(random_state=54321, n_estimators=40)
model_RFC.fit(train_features, train_target)
score_rfc = model_RFC.score(test_features, test_target)

print(f"A acurácia do modelo Logistic Regression para o conjunto de teste foi igual a: {score_lr}")
print(f"A acurácia do modelo Random Forest Classifier para o conjunto de teste foi igual a: {score_rfc}")

A acurácia do modelo Logistic Regression para o conjunto de teste foi igual a: 0.7776049766718507
A acurácia do modelo Random Forest Classifier para o conjunto de teste foi igual a: 0.8242612752721618


### Resultados

Considerando os dados entregues, foram avaliados dois modelos de classificação: Logistic Regression e Random Forest Classifier. Ambos os modelos se adequaram ao piso mínimo estabelecido de 75% de acurácia, porém o segundo obteve melhor resultado, atingindo 82% de acurácia referente ao conjunto de dados de teste determinado.

Desta forma, para o problema proposto e para o conjunto de dados entregue, o modelo que melhor consegue prever o plano a qual determinado perfil de cliente irá aderir é o Random Forest Classifier.